### PR TITLE
Support simplified sign and notarization for macOS distribution in Gradle plugin

### LINKED ISSUE #331: Support simplified sign and notarization for macOS distribution in Gradle plugin
Try to sign an example, fix issues and add a tutorial.

### PR DESCRIPTION
**Pull request by AlexeyTsvetkov**
Fixes #331

### CODE REVIEW

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/dsl/PlatformSettings.kt [hunk 1]
```diff
@@ -3,34 +3,65 @@ package org.jetbrains.compose.desktop.application.dsl
...
+    var bundleID: String? = null
+
+    internal var signSettings: MacOSSigningSettings? = null
     fun signing(fn: Action<MacOSSigningSettings>) {
-        // enable sign if it the corresponding block is present in DSL
-        if (!isSignInitialized) {
-            isSignInitialized = true
-            signing.sign = true
+        check(signSettings == null) { "Signing is already configured" }
+        signSettings = MacOSSigningSettings().also {
```
Comment (olonho) [inline gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/dsl/PlatformSettings.kt:29]: Maybe `signingSettings` for symmetry?

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/MacJarSignFileCopyingProcessor.kt [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+) : FileCopyingProcessor {
+    private val bundleId = macBundleID.orNull
+    private val keychainPath = signSettings.keychain
+    private val signPrefix: String
+    private val signKey: String
+
+    init {
+        check(currentOS == OS.MacOS) { "$currentOS is not compatible with ${this::class.java}" }
+        check(bundleId != null) {
+            "Signing requires to specify unique application's identifier  (e.g. com.mycompany.myapp). Specify an identifier using DSL like so:\n" +
```
Comment (olonho) [inline gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/MacJarSignFileCopyingProcessor.kt:26]: Why not single multiline literal?

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/MacJarSignFileCopyingProcessor.kt [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+            """|nativeExecutables {
+               |  macOS {
+               |      bundleID = "com.mycompany.myapp"
+               |  }
+               |bundleID may only contain alphanumeric characters (A-Z,a-z,0-9), hyphen (-) and period (.) characters}
+               |""".trimMargin()
+        }
+        check(bundleId.matches("[A-Za-z0-9\\-\\.]+".toRegex())) {
```
Comment (olonho) [inline gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/MacJarSignFileCopyingProcessor.kt:34]: Likely number, hyphen and dot cannot be the first character?

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/MacJarSignFileCopyingProcessor.kt [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        }
+        signPrefix = signSettings.signPrefix
+            ?: (bundleId.substringBeforeLast(".") + ".").takeIf { bundleId.contains('.') }
+            ?: error("Could not infer 'signPrefix'. Specify explicitly or use reverse DNS notation for bundleID")
+
+        val identity = signSettings.identity
+
+        val developerIdPrefix = "Developer ID Application: "
+        val thirdPartyMacDeveloperPrefix = "3rd Party Mac Developer Application: "
+        val signIdentity = when {
```
Comment (olonho) [inline gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/MacJarSignFileCopyingProcessor.kt:46]: Nice.

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/MacJarSignFileCopyingProcessor.kt [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+                    exec.executable = MacUtils.security.absolutePath
+                    val args = arrayListOf("find-certificate", "-a", "-c", signIdentity)
+                    keychainPath?.let { args.add(it) }
+                    exec.args(*args.toTypedArray())
+                    exec.standardOutput = ps
+                }
+            }
+            baos.toString()
+        }
+        val regex = Pattern.compile("\"alis\"<blob>=\"([^\"]+)\"")
```
Comment (olonho) [inline gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/MacJarSignFileCopyingProcessor.kt:64]: What is `alis`?

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/MacJarSignFileCopyingProcessor.kt [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        }
+    }
+
+    private fun copyAndSignNativeLibs(zin: ZipInputStream, zout: ZipOutputStream) {
+        for (sourceEntry in generateSequence { zin.nextEntry }) {
+            if (!sourceEntry.name.endsWith(".dylib")) {
+                zout.putNextEntry(ZipEntry(sourceEntry))
+                zin.copyTo(zout)
+            } else {
+
```
Comment (olonho) [inline gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/MacJarSignFileCopyingProcessor.kt:98]: Extra line.

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/tasks/AbstractUploadAppForNotarizationTask.kt [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        val (res, output) = ByteArrayOutputStream().use { baos ->
+            PrintStream(baos).use { ps ->
+                val res = execOperations.exec { exec ->
+                    exec.executable = MacUtils.xcrun.absolutePath
+                    exec.args(
+                        "altool",
+                        "--notarize-app",
+                        "--primary-bundle-id", bundleId,
+                        "--username", notarization.appleID,
+                        "--password", notarization.password,
```
Comment (olonho) [inline gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/tasks/AbstractUploadAppForNotarizationTask.kt:68]: Shall we check that appleID and password are not empty?

File: tutorials/Signing_and_notarization_on_macOS/README.md [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+If you have multiple `Developer ID Application` certificates installed,
+you will need to specify the path to the keychain, containing
+the certificate intended for signing.
+
+## Preparing an App ID
+
+#### Viewing existing App IDs
+
+Open https://developer.apple.com/account/resources/identifiers/list
```
Comment (olonho) [inline tutorials/Signing_and_notarization_on_macOS/README.md:56]: Markdown supports hyperlinks.

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/MacJarSignFileCopyingProcessor.kt (hunk 5 above)
Comment (olonho) [reply gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/MacJarSignFileCopyingProcessor.kt:64]: I see, `find-certificate` uses four letter abbreviations.

File: tutorials/Signing_and_notarization_on_macOS/README.md [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+             --password <password>
+```
+
+Then you'll be able to refer to the password like `@keychain:NOTARIZATION_PASSWORD`
+without the need to write the password itself.
+
+## Configuring Gradle
+
+All properties are set in `macOS` DSL block of Compose Desktop DSL:
+```kotlin
```
Comment (olonho) [inline tutorials/Signing_and_notarization_on_macOS/README.md:97]: Use "``` kotlin" note space to mark code snippets that do not contain self-sufficient Kotlin code for tutorials compilability checker.

Comment (olonho) [reply tutorials/Signing_and_notarization_on_macOS/README.md:97]: See https://github.com/JetBrains/compose-jb/blob/ff73f8268a972f13cb056e71ef7420ab2f43d3ff/tutorials/checker/build.gradle.kts#L23

File: tutorials/Signing_and_notarization_on_macOS/README.md [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+            macOS {
+                bundleID = "com.example-company.example-app"
+            
+                signing { 
+                   identity = "John Doe"
+                }
+                notarization {
+                   appleID = "john.doe@example-company.com"
+                   password = "@keychain:NOTARIZATION_PASSWORD"
```
Comment (olonho) [inline tutorials/Signing_and_notarization_on_macOS/README.md:125]: Suggest people adding password and Apple ID to buildscript doesn't look like a good idea. Consider getting those from Gradle properties.

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/MacJarSignFileCopyingProcessor.kt (hunk 3 above)
Comment (AlexeyTsvetkov) [reply gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/MacJarSignFileCopyingProcessor.kt:34]: I don't see such restrictions in Apple's documentation: https://developer.apple.com/documentation/bundleresources/information_property_list/cfbundleidentifier

**Review by olonho — DISMISSED** [review]

**Review by olonho — APPROVED** [review]

**Pull request merged**